# STATS Tutorial Homework

TUTORIAL TRANSCRIPT https://chatgpt.com/share/7f23c00b-a60b-4918-b77e-4e08d7b33589

HOMEWORK TRANSCRIPT https://chatgpt.com/share/0efdec9f-3b6f-47c7-a250-fe695640168e

In [2]:
import pandas as pd

# URL of the dataset
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'

# Load the dataset into a DataFrame
df = pd.read_csv(url)
print(df)

     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ... 

In [3]:
df["num missing"] = df.isna().s sum(axis=1)
df


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  num missing  
0        0         A/5 21171   7.2500   NaN        S            1  
1        0          PC 17599  71.2833   C85        C            0  
2        0  STON/O2. 3101282   7.9250   NaN        S            1  
3        0            113803  53.1000  C123        S            0  
4        0            373450   8.0500   NaN        S            1  
..     ...               ...      ...   ...      ...          ...  
886      0            211536  13.0000   NaN        S            1  
887      0            112053  30.0000   B42        S            0  
888      2        W./C. 6607  23.4500   NaN        S            2  
889      0            111369  30.0000  C148        C            0  
890      0            370376   7.7500   NaN        Q            1  

[891 rows x 13 columns]

In [5]:
df.dropna()[:15][['Name','Survived']]

     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ... 

Name  Survived
1   Cumings, Mrs. John Bradley (Florence Briggs Th...         1
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)         1
6                             McCarthy, Mr. Timothy J         0
10                    Sandstrom, Miss. Marguerite Rut         1
11                           Bonnell, Miss. Elizabeth         1
21                              Beesley, Mr. Lawrence         1
23                       Sloper, Mr. William Thompson         1
27                     Fortune, Mr. Charles Alexander         0
52           Harper, Mrs. Henry Sleeper (Myna Haxtun)         1
54                     Ostby, Mr. Engelhart Cornelius         0
62                        Harris, Mr. Henry Birkhardt         0
66                       Nye, Mrs. (Elizabeth Ramell)         1
75                            Moen, Mr. Sigurd Hansen         0
88                         Fortune, Miss. Mabel Helen         1
92                        Chaffee, Mr. Herbert Fuller         0

In [6]:
# Get the number of rows and columns
rows, columns = df.shape

# Print the results
print(f'The data frame has {rows} rows and {columns} columns.')


The data frame has 891 rows and 13 columns.


As stated by ChatGPT, but reworded into my own thinking, the way that the observations and variables work in the context of this dataset is as follows;

A variable represents the entire dataset, but on one specific topic, for example, the passenger's name, or their class.

An observation represents one specific passenger, whether it be Mrs Jacques Heath on row 3 or Mr Timothy J on Row 6. It is the combination of all the variables, but all referencing a single individual.

In [7]:
# Summary of numeric columns
print(df.describe())

# Summary of all columns (including non-numeric)
print(df.describe(include='all'))


       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  num missing  
count  891.000000  891.000000   891.000000  
mean     0.381594   32.204208     0.971942  
std      0.806057   49.693429     0.618351  
min      0.000000    0.000000     0.000000  
25%      0.000000    7.910400     1.000000  
50%      0.000000   14.454200     1.000000  
75%      0.000000   31.000000     1.000000  
max      

It can be noted that in a handful of variables, the count is severely under the 891 expected for one per observation. This is due to null values being omitted. In this case, if there is missing data, the observation will be omitted when calculating various aspects like the mean, std, min, etc and the count will be lowered.  

Another thing to note is that the first iteration of this code will only take numeric values into calculation, whereas the second iteration will take ALL data, including non numeric. For numerical values, we should utilize the top one, however, for information like the number of males compared to females, which was provided in the form of "male" "female" statistics (strings), you can look at the "include='all'" iteration of the code.

The difference between an attribute and a method is as follows;

- An attribute is effectively data that is already in the object, in a reference to Python Classes, it would be something that requires no calculation and can be obtained by simply getting the variable name.

- A method is a function, it is data that requires calculation and can potentially have parameters passed through it. For example, in the above code, df.describe() uses an argument of "include='all'", which tells the method to also include the non numeric data and work with it as though it were numeric data.

# POST LECTURE HOMEWORK

#6 - The df.describe() method provides the 'count','mean','std','min',etc summary for each variable;

- Count represents the number of items accounted for in the calculations below. It will omit missing values as they are not used to calculate any of the data below.

#7 - dropna() will delete any rows which may have missing data, this is useful if you require all of your data to be complete, or if you're doing a stats assignment where your professor wants you to work without missing data.

However, if a lot of the data is missing in a handful of columns, you can del df['col'] to delete an entire column in question. For example in the titanic dataset, it is shown that there are only 204 observations where cabin is not n/a. If we dropna'd, all but 204 rows would be deleted, drastically reducing the dataset size. In this case, if we don't need the cabin data, we can del df['col'] to just remove the Cabin variable.

In another scenario, if we want to remove all N/A rows, EXCEPT for those which are exclusively N/A in the Age category, we can delete the Age category first, then dropna. This is just one use case where it may allow you to preserve more data than dropna() while still allowing you to have a complete dataset.

In the next line of the code, I will be dropping the Cabin line of code, then dropna'ing. This will remove any observations that are missing information other than the cabin number. This is because the cabin variable is already incredibly limited, only having 200 observations, as such, dropna'ing with it still in the dataset will remove all but 200 of our rows.

In [12]:
del df['Cabin']
df = df.dropna()

df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked  num missing  
0        0         A/5 21171   7.2500        S            1  
1        0          PC 17599  71.2833        C            0  
2        0  STON/O2. 3101282   7.9250        S            1  
3        0            113803  53.1000        S            0  
4        0            373450   8.0500        S            1  
..     ...               ...      ...      ...          ...  
885      5            382652  29.1250        Q            1  
886      0            211536  13.0000        S            1  
887      0            112053  30.0000        S            0  
889      0            111369  30.0000        C            0  
890      0            370376   7.7500        Q            1  

[712 rows x 12 columns]

This is question #8, and below is the import for the dataset. Please refresh the kernel before running this.

In [1]:
import pandas as pd

# URL for the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Download and load the dataset into a pandas DataFrame
df = pd.read_csv(url)

# Display the first few rows of the dataset
print(df.head())

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  


ChatGPT broke down the groupby expression, but did not give an example. The following will be my usage of the entire groupby()[].describe() function.

In [3]:
df.groupby("survived")["fare"].describe()

count       mean        std  min      25%   50%   75%       max
survived                                                                 
0         549.0  22.117887  31.388207  0.0   7.8542  10.5  26.0  263.0000
1         342.0  48.395408  66.596998  0.0  12.4750  26.0  57.0  512.3292

Not at all related to the assignment, but the average person who survived paid over twice as much on their fare as those who did not survive. This makes a lot of sense, as the first class passengers of the Titanic were likely the first ones to get access to a lifeboat.

2. The count line of the data in this case includes null values. In the df.groupby expression, all null values are still accounted for (but just ignored). However, in the following line, I will run the same code after running df.dropna().

In [4]:
df = df.dropna()
df.groupby("survived")["fare"].describe()

count       mean        std   min    25%      50%   75%       max
survived                                                                   
0          59.0  64.532131  62.108090  0.00  27.15  50.4958  79.2  263.0000
1         123.0  85.821107  81.843522  8.05  30.50  69.3000  93.5  512.3292

As there is significantly less data (most of the data was dropped due to missing values), the count is significantly lower, as well as the rest of the data. It is shown here that generally, the richer passengers were more reported on, and as such had less missing data.

In the next line of the code, I will intentionally implement errors into the code as requested by question #3.

In [1]:
# URL for the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Download and load the dataset into a pandas DataFrame
df = pd.read_csv(url)

# Display the first few rows of the dataset
print(df.head())

NameError: name 'pd' is not defined

In [2]:
import pandas as pd

# URL for the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"

# Download and load the dataset into a pandas DataFrame
df = pd.read_csv(url)

# Display the first few rows of the dataset
print(df.head())


HTTPError: HTTP Error 404: Not Found

In [3]:
import pandas as pd

# URL for the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Download and load the dataset into a pandas DataFrame
df = pd.read_csv(url)

# Display the first few rows of the dataset
print(DF.head())


NameError: name 'DF' is not defined

In [4]:
import pandas as pd

# URL for the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Download and load the dataset into a pandas DataFrame
df = pd.read_csv(url

# Display the first few rows of the dataset
print(df.head())


SyntaxError: '(' was never closed (344142628.py, line 7)

In [5]:
import pandas as pd

# URL for the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Download and load the dataset into a pandas DataFrame
df = pd.read_csv(url)

# Display the first few rows of the dataset
print(df.hed())


AttributeError: 'DataFrame' object has no attribute 'hed'

In [6]:
import pandas as pd

# URL for the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Download and load the dataset into a pandas DataFrame
df = pd.read_csv(url)

# Display the first few rows of the dataset
print(df.head())

df.groupby("Age")["class"].describe()

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  


KeyError: 'Age'

In [7]:
import pandas as pd

# URL for the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Download and load the dataset into a pandas DataFrame
df = pd.read_csv(url)

# Display the first few rows of the dataset
print(df.head())

df.groupby("age")[pclass].describe()

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  


NameError: name 'pclass' is not defined

In [8]:
import pandas as pd

# URL for the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Download and load the dataset into a pandas DataFrame
df = pd.read_csv(url)

# Display the first few rows of the dataset
print(df.head())

# Group by 'age' and describe 'pclass'
grouped = df.groupby("age")["pclass"].describe()
print(grouped)


   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
       count  mean       std  min   25%  50%   75%  max
age                                                    
0.42     1.0   3.0       NaN  3.0  3.00  3.0  3.00  3.0
0.67     1.0   2.0       NaN  2.0  2.00  2.0  2.00 

A google search is significantly easier than asking chatgpt. Additionally, a lot of these errors could be deciphered purely off of the error message itself, for stuff such as "missing parenthesis", etc. I also started feeling bad for chatgpt as it kept giving me the correct code and I had to keep messing it up to send it back.

https://chatgpt.com/share/6e270a4f-d4c0-46b9-a799-53503d32f821 

9. Somewhat? Perhaps?